### PR TITLE
feat(meta): implement create sink

### PR DESCRIPTION
**Pull request by nanderstabel**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR should implement CREATE SINK feature.

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Frontend now produces a `StreamFragmentGraph` which consists of an upstream `ChainNode` and a downstream `SinkNode` (`src/frontend/src/handler/create_sink.rs`).
- Frontend implements methods to add and delete sink to/from `SchemaCatalog`
- Meta receives the `CreateSinkRequest` from frontend. Dependent relations are checked + `SinkManager` is initialized.
- Via the `SinkManager` a create sink request is send to the Compute node. 
- Compute node can add/remove `Sink` to/from `MemSinkManager` (located in `src/sink` repository).

`./e2e_test/sink/basic_test.slt` is runnable, however the only things that currently happens is that sink is added/removed to from the respective nodes and the communication between frontend/meta/compute. Basically the framework is there, but the actual implementation of create sink is still lacking. 

Crucial missing functionalities:
- Meta: Actual actors need to be generated using `ActorGraphBuilder` and "connected" to the upstream materialized view.
- Compute: `SinkExecutorBuilder` needs to be called and executed.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Connector (sources & sinks)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

**Review by neverchanje — APPROVED** [review]
Overall I have no problem on this PR after a thorough read. I think we can merge it as long as the CI passes.

**Comment by neverchanje** [comment]
Eventually, we can implement a fake sink that collects all the messages in memory so that we might be able to verify the execution path in unit tests.

File: src/meta/src/stream/sink_manager.rs
Comment (yezizp2012) [inline src/meta/src/stream/sink_manager.rs:162]: IIUC sink manager does not have to care about some change in downstream like redis/mysql etc. So it don't have to  monitor any changes.

File: src/sink/src/manager.rs
Comment (yezizp2012) [inline src/sink/src/manager.rs:33]: Ditto, local sink manger in CN is not necessary.

**Comment by tabVersion** [comment]
please hold the pr for a while, it is too long to review. 

next time we can use an issue to track progress instead of implementing all in one pr

File: e2e_test/sink/basic_test.slt
Comment (tabVersion) [inline e2e_test/sink/basic_test.slt:8]: `'kafka.scan.startup.mode' = 'earliest'` is unnecessary

**Comment by shanicky** [comment]
Unlike source, a sink is not reused and does not require a long-running thread or coroutine to discover new partitions, so a sink manager on either meta node or compute node may not be necessary

File: src/meta/src/stream/sink_manager.rs
Comment (tabVersion) [inline src/meta/src/stream/sink_manager.rs:166]: same here. I think meta part does not need to deal with downstream partition changes since broker can help with it   .

**Comment by neverchanje** [comment]
> Unlike source, a sink is not reused and does not require a long-running thread or coroutine to discover new partitions, so a sink manager on either meta node or compute node may not be necessary

I'm not 100% sure about it, do we need long-running threads for Kafka sink? @tabVersion

**Comment by tabVersion** [comment]
> > Unlike source, a sink is not reused and does not require a long-running thread or coroutine to discover new partitions, so a sink manager on either meta node or compute node may not be necessary
> 
> I'm not 100% sure about it, do we need long-running threads for Kafka sink? @tabVersion

I think we don't need that. we can use message key to guarantee messages in one epoch and one actor to be on the same partition.

**Comment by nanderstabel** [comment]
> Let's merge asap and fix things anytime needed

@neverchanje Are you able to force merge it?

**Comment by tabVersion** [comment]
> > Let's merge asap and fix things anytime needed
> 
> @neverchanje Are you able to force merge it?

@nanderstabel please add some doc in the release note part if is there is any required parameter within `with` clause.

File: src/meta/src/rpc/service/ddl_service.rs
Comment (BugenZhao) [inline src/meta/src/rpc/service/ddl_service.rs:248]: What is `mview` here?

File: src/meta/src/rpc/service/ddl_service.rs
Comment (BugenZhao) [inline src/meta/src/rpc/service/ddl_service.rs:259]: Can't we reuse the code instead of duplicating them?

File: src/meta/src/rpc/service/ddl_service.rs
Comment (BugenZhao) [inline src/meta/src/rpc/service/ddl_service.rs:357]: Clean-up unused code?

File: src/meta/src/stream/stream_manager.rs
Comment (BugenZhao) [inline src/meta/src/stream/stream_manager.rs:689]: Please avoid duplicating so many codes.

**Review by BugenZhao — CHANGES_REQUESTED** [review]

File: src/meta/src/rpc/service/ddl_service.rs
Comment (nanderstabel) [reply src/meta/src/rpc/service/ddl_service.rs]: `mview` here in `create_sink` is a `SInk` so I've renamed it as such.

File: src/meta/src/rpc/service/ddl_service.rs
Comment (nanderstabel) [reply src/meta/src/rpc/service/ddl_service.rs:257]: I have introduced `fn get_dependent_relations(&self, fragment_graph: &StreamFragmentGraph) -> RwResult<Vec<u32>>`  method to reduce code duplication in `fn create_sink` and `fn create_materialized_view`.

File: src/meta/src/rpc/service/ddl_service.rs
Comment (nanderstabel) [reply src/meta/src/rpc/service/ddl_service.rs]: I guess this was leftover code from copying catalog implementation in `create_materialized_view`. I've replaced it with the corresponding catalog functionality for Sink.

File: src/meta/src/stream/stream_manager.rs
Comment (nanderstabel) [reply src/meta/src/stream/stream_manager.rs:689]: Apart from `fn get_dependent_relations` I also introduced the `fn get_internal_table(&self, ctx: &CreateMaterializedViewContext) -> RwResult<Vec<Table>>` method to reduce code duplication in `fn create_sink`, `create_materialized_view` and `create_materialized_source_inner`.

**Comment by nanderstabel** [comment]
Thanks to all of your help and feedback it now looks like `SinkExecutor` is able to run now.

**Comment by tabVersion** [comment]
image
Looks like sink executor is handled well in dep graph

**Comment by skyzh** [comment]
Generally looks good to me, but it seems that we're duplicating too much code from creating MV. This would make our code hard to maintain. Shall we hold this PR a little bit and do some refactors before merging?

**Comment by skyzh** [comment]
From my perspective, create sink should be yet another a kind of create MV. May refer to the implementation of `create index` and see how we can optimize this.

**Comment by skyzh** [comment]
image

just did a quick diff between create_materialized_view and create_sink. The body is totally the same. We can simply call `create_materialized_view` in `create_sink`.

**Comment by shanicky** [comment]
> Generally looks good to me, but it seems that we're duplicating too much code from creating MV. This would make our code hard to maintain. Shall we hold this PR a little bit and do some refactors before merging?

This PR should not be able to be merged yet, some modifications are still needed

**Comment by tabVersion** [comment]
> > Generally looks good to me, but it seems that we're duplicating too much code from creating MV. This would make our code hard to maintain. Shall we hold this PR a little bit and do some refactors before merging?
> 
> This PR should not be able to be merged yet, some modifications are still needed

I think the pr is good enough now. Let's re-review it and merge asap @shanicky @BugenZhao

File: src/stream/src/executor/sink.rs
Comment (BugenZhao) [inline src/stream/src/executor/sink.rs:70]: Will this be removed after we adopt the new transaction interfaces? cc @tabVersion

File: src/meta/src/rpc/service/ddl_service.rs
Comment (BugenZhao) [inline src/meta/src/rpc/service/ddl_service.rs:526]:
```suggestion
        let relation_id = match relation {
```

File: src/meta/src/rpc/service/ddl_service.rs
Comment (BugenZhao) [inline src/meta/src/rpc/service/ddl_service.rs:390]: How about making this function static? Seems `self` is unused.
```suggestion
    fn get_dependent_relations(fragment_graph: &StreamFragmentGraph) -> RwResult<Vec<TableId>> {
```

File: src/meta/src/rpc/service/ddl_service.rs
Comment (BugenZhao) [inline src/meta/src/rpc/service/ddl_service.rs:419]: Ditto.

File: src/meta/src/rpc/server.rs
Comment (BugenZhao) [inline src/meta/src/rpc/server.rs:408]:
```suggestion
```

File: src/meta/src/manager/relation.rs
Comment (BugenZhao) [inline src/meta/src/manager/relation.rs:21]: cc @st1page, does this design LGTY?

File: e2e_test/batch/basic/temp.slt.part
Comment (BugenZhao) [inline e2e_test/batch/basic/temp.slt.part:1]: What's this file for?

**Review by BugenZhao — COMMENTED** [review]
Good work! Much clearer than before.

File: src/sink/Cargo.toml
Comment (skyzh) [inline src/sink/Cargo.toml:16]: Add workspace_hack here?

File: e2e_test/sink/aggregation.slt
Comment (skyzh) [inline e2e_test/sink/aggregation.slt:1]: Is there any plan to run this in CI?

File: src/compute/src/server.rs
Comment (skyzh) [inline src/compute/src/server.rs:94]: Why is it unused for now?

File: src/sink/Cargo.toml
Comment (skyzh) [inline src/sink/Cargo.toml:4]: should be consistent with other crates (0.1.10)

**Review by skyzh — COMMENTED** [review]
* The sink crate doesn't contain any sink. Is this expected?
* risingwave_sink is not added into root Cargo.toml as workspace member. May add it.

File: e2e_test/sink/aggregation.slt
Comment (skyzh) [inline e2e_test/sink/aggregation.slt:8]: What does `append_only` mean here? The output from HashAgg doesn't seem to be append-only...

**Comment by skyzh** [comment]
Panicked when trying to manually run something similar the e2e test.

```
thread 'tokio-runtime-worker' panicked at 'no entry found for key', src/meta/src/stream/stream_manager.rs:170:21
```

Trying to reproduce...

**Comment by skyzh** [comment]
```
create table t1(v1 int);
create materialized view mv1 as select sum(v1) as sum_v1 from t1;
CREATE SINK si2 FROM mv1 WITH (kafka.brokers = '127.0.0.1:29092', kafka.topic = 'test_topic_2', sink.type = 'append_only', sink_type = 'kafka');
```

will instantly panic.

**Comment by skyzh** [comment]
I think that's another bug. I'll submit a separate issue.

**Comment by skyzh** [comment]
related: https://github.com/singularity-data/risingwave/issues/4067

File: src/meta/src/manager/relation.rs
Comment (st1page) [reply src/meta/src/manager/relation.rs:21]: No, I think here the code re-using is for creating the streaming job on the Compute Node. So the "Relation" here seems just a stream destination job.

**Comment by tabVersion** [comment]
> * The sink crate doesn't contain any sink. Is this expected?

I think it's not expected. 

> * risingwave_sink is not added into root Cargo.toml as workspace member. May add it.

I am thinking about removing the crate. Let's hold for now

File: src/meta/src/rpc/service/ddl_service.rs
Comment (nanderstabel) [reply src/meta/src/rpc/service/ddl_service.rs:390]: I agree, just fixed it 👍

File: src/meta/src/rpc/service/ddl_service.rs
Comment (nanderstabel) [reply src/meta/src/rpc/service/ddl_service.rs:419]: `fn get_internal_table` needs `self` to access `fn set_table_mapping`. I could change it's signature by adding an argument but that seems unnecessary so I rather not touch it.

File: e2e_test/batch/basic/temp.slt.part
Comment (nanderstabel) [reply e2e_test/batch/basic/temp.slt.part:1]: Slipped in there by accident. Removed it 👍

File: src/meta/src/manager/relation.rs
Comment (nanderstabel) [reply src/meta/src/manager/relation.rs:21]: @yezizp2012 I did consider adding `Source`, but for another reason. `Relation` can also be used to reduce code duplication in [CatalogManager](https://github.com/singularity-data/risingwave/blob/main/src/meta/src/manager/catalog.rs). `fn start_create_[source/sink/table]_procedure()` and `fn cancel_create_[source/sink/table]_procedure()` since these methods are practically the same. 

@st1page for me to be sure, are you ok with this design?

File: e2e_test/sink/aggregation.slt
Comment (nanderstabel) [reply e2e_test/sink/aggregation.slt:1]: This test file slipped in there by accident. I used it to fix some unexpected behavior. I see you've already created an issue for it (https://github.com/singularity-data/risingwave/issues/4067) thanks!.

**Comment by nanderstabel** [comment]
> 

Since a lot of code in the original version of this PR turned out to be obsolete, this sink crate is now redundant and not used anywhere. I think it should just be deleted, at least for now?

**Comment by nanderstabel** [comment]
> > * The sink crate doesn't contain any sink. Is this expected?
> 
> I think it's not expected.
> 
> > * risingwave_sink is not added into root Cargo.toml as workspace member. May add it.
> 
> I am thinking about removing the crate. Let's hold for now

Since a lot of code in the original version of this PR turned out to be obsolete, this sink crate is now redundant and not used anywhere. I think it should just be deleted, at least for now?

File: src/meta/src/manager/id.rs
Comment (yezizp2012) [inline src/meta/src/manager/id.rs:180]: Since `create_sink` is using `Table` category and reusing code of creating materialized view, the fragment graph will be stored as `TableFragment` in `FragmentManager`. The key in manager is `SinkId` and `TableId`.  That's acceptable and this sink category is never used and can be removed.

File: src/connector/src/sink/kafka.rs
Comment (neverchanje) [inline src/connector/src/sink/kafka.rs:95]: Need comment on `u64`.

File: src/connector/src/sink/kafka.rs
Comment (neverchanje) [inline src/connector/src/sink/kafka.rs:101]:
```suggestion
    state: KafkaSinkState,
```

File: src/meta/src/manager/catalog.rs
Comment (neverchanje) [inline src/meta/src/manager/catalog.rs:715]:
```suggestion
            bail!(
```

File: src/frontend/src/test_utils.rs
Comment (nanderstabel) [reply src/frontend/src/test_utils.rs:344]: Probably not since this method is only used for testing in the frontend. Same goes for:
```
    async fn create_materialized_view(
        &self,
        mut table: ProstTable,
        _graph: StreamFragmentGraph,
```

File: src/meta/src/manager/catalog.rs
Comment (nanderstabel) [reply src/meta/src/manager/catalog.rs:715]: That's a nice macro 💯

**Pull request merged**
